### ISSUE TITLE
[feature proposal] Option for removing IL bodies from R2R images

### ISSUE THREAD

**Issue by raffaeler** [dotnet/runtime#52190]
For years, the native vs managed debate was centered around the intellectual property / obfuscation problem.
I still periodically discuss with customers about the need to make it harder the decompilation of certain parts of an application.
I am very aware that any language can be reversed and therefore this is just a matter of rising the bar.
Anyway, talking about .NET adoption, the obfuscation problem is often brought as an example.

In the current versions, .NET is now able to perform AOT on various bases, but still keeping the IL Body of the methods inside the assemblies. In certain cases the runtime still needs them (generics is just an example), but in others the IL could probably be removed without any impact.

I am wondering whether it would be possible, for limited portions of an assembly to:
- leave all the metadata (to avoid reflection issues and be able to reference it)
- perform AOT (by targeting a specific platform with a RID)
- remove certain IL Bodies [it would be helpful for the user to tag (with an attribute) the desired ones, so that we could now from the compiler whether there was some reason which prevented it]

In addition to the "obfuscation", there could be also a significant gain in terms of assembly size reduction.

I already looked at the CoreRT solution, but this works only if you want to generate a "C" style LIB or fully AOT the entire solution, which is not feasible for many applications and of course libraries.

Thoughts? Comments?

**Comment by Symbai** [comment]
> Thoughts? Comments?

Yes, the .NET team confirmed several times already that the AOT they have is all they deliver and there are no plans for improvements (i.e. making it work for everything else than libraries or console projects). There are even no plans on moving it out of the experimental so far. And there was a survey quite some time ago and obfuscation was ranked very badly while (startup) performance was one of the 1#. Which means removing IL body isn't a big thing (I dont like it either but it is as it is). And regarding our last statement, WPF is dead and won't even get trimming support. Winforms isn't dead but out of the .NET team scope as well. If we're "lucky" the most we might get, in 5 years, is Xamarin / Maui / WinUI AOT support.

**Comment by benaadams** [comment]
> there are no plans for improvements

There are 3 variants of AoT being worked on in runtimelab https://github.com/dotnet/runtimelab

**Comment by raffaeler** [comment]
@benaadams Yes, I know. The folks there suggested me to post here

**Comment by mangod9** [comment]
Thanks for opening a feature request. In the current model of R2R, pre-compiled code is treated as a "cache" of native code and IL is still required to potentially generate better quality code with Tiering/Dynamic PGO etc. But the request is certainly valid for certain scenarios and we will investigate for .net 6+

**Comment by raffaeler** [comment]
Thank you @mangod9 
I am familiar with tiered compilation and I guess it should not be too tricky for R2R not to try regenerating the code if the IL has been removed.

I am wondering instead what are the challenges in informing the user for a method the user requested to remove but instead is needed by the runtime and cannot be removed.
Maybe just reporting the methods that cannot be removed could be enough.

When you say .net 6+, does this mean that is too late for v6 or there is still the chance to fit it?

**Comment by mangod9** [comment]
> When you say .net 6+, does this mean that is too late for v6 or there is still the chance to fit it?

Correct, .net 6 feature work is expected to complete in the next couple of months so unlikely we have space for any new functionality.

**Comment by raffaeler** [comment]
@mangod9 @jkotas Sorry for the direct ping. Any news on this?
Thanks!

**Comment by mangod9** [comment]
adding @agocke who might be able to determine feasibility in 11.

**Comment by agocke** [comment]
This is against the design goals of crossgen2. R2R code is meant to be a performance optimization only -- existing applications should not stop working or change behavior. That means that IL needs to be present, both for features like reflection to work properly, and for the JIT to have the opportunity to re-JIT if the performance metrics or environmental features require it.

Removing IL fundamentally changes the behavior of applications because certain features, like reflection and assembly loading, stop working. Those things are possible in Native AOT because Native AOT explicitly has restrictions that regular .NET does not, but we have no plans to bring those restrictions to R2R.

**Comment by agocke** [comment]
To respond to a specific part of the proposal:

>  am wondering whether it would be possible, for limited portions of an assembly to:
> leave all the metadata (to avoid reflection issues and be able to reference it)
> perform AOT (by targeting a specific platform with a RID)
> remove certain IL Bodies [it would be helpful for the user to tag (with an attribute) the desired ones, so that we could now from the compiler whether there was some reason which prevented it]

I'm not opposed to the general idea of removing IL when it's possible to guarantee no behavioral changes, but I think that's a perf-driven discusison, not obfuscation. For reliable removal of IL, Native AOT is our answer. Any feature around removing IL for performance will be based on whether it's the right tradeoff. Right now it's not planned, but that may change.
